### PR TITLE
Tomer/training_artifacts

### PR DESCRIPTION
**Pull request by Z30G0D**
draft for training artifacts
looks something like this:
[report](https://wandb.ai/deep_voice/finding_willy/reports/Artifacts-uplaod-report--VmlldzoxNTgzMTYw?accessToken=by47bhpngjfmwz7nrl6k52d9n51nd0uc5271ei7xr10uw091djvqa1y5wiosgx2t)

Run is [here](https://wandb.ai/deep_voice/finding_willy/runs/2nnuxaun/overview?workspace=user-deep_voice)

### CODE REVIEW

File: soundbay/utils/logging.py [hunk 1]
```diff
@@ -125,11 +128,43 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+    def upload_artifacts(self, audio: torch.Tensor, label: torch.Tensor, raw_wav: torch.Tensor, idx: torch.Tensor, sample_rate: int=16000, flag: str='train', epoch: int=1):
+        """upload algorithm artifacts to W&B during training session"""
+        matplotlib.use('Agg')
+        idx = idx.detach().cpu().numpy()
+        label = label.detach().cpu().numpy()
+
+
+        # Original wavs batch
+
+        artifact_wav = torch.squeeze(raw_wav).detach().cpu().numpy() * 100
```
Comment (noamzbr) [inline soundbay/utils/logging.py:145]: Think there could be a simple solution for normalization here - bring audio power up to "volume" dB, with X an argument.

This can be something like

```
artifact_wav = torch.squeeze(raw_wav).detach().cpu().numpy()
artifact_wav = artifact_wav * 10 ^ (volume / (20 * log10(np.std(artifact_wav ))))
```

File: soundbay/utils/signal_processing.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import torch
 
 
+
```
Comment (noamzbr) [inline soundbay/utils/signal_processing.py:6]: think this is here by mistake

File: soundbay/utils/logging.py [hunk 3]
```diff
@@ -125,11 +128,43 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+        list_of_specs_objects = [wandb.Image(data_or_path=spec, caption=f'{ind}_train') for spec, ind in zip(specs,idx)]
+        log_dict1 = {f'First batch {flag} original wavs for epoch {epoch}': list_of_wavs_objects}
+        log_dict2 = {f'First batch {flag} augmented spectrograms for epoch {epoch}': list_of_specs_objects}
+
+        # Upload to W&B
+        wandb.log(log_dict1, commit=True)
+        wandb.log(log_dict2, commit=True)
+
+        # Clear figures
+        plt.figure().clear()
```
Comment (noamzbr) [inline soundbay/utils/logging.py:162]: not critical, but would be nice to know which one of these is actually required(:

File: soundbay/utils/signal_processing.py [hunk 4]
```diff
@@ -33,6 +34,8 @@ def __call__(self,  sample):
         return melspectrogram
 
 
+
```
Comment (noamzbr) [inline soundbay/utils/signal_processing.py:37]: same

**Review by noamzbr — COMMENTED** [review]
Looks good! Most of my comments are on stuff left over from debugging

**Review by noamzbr — COMMENTED** [review]
In the future we should consider adding time and frequency axis to the plots, but it can defintly wait

File: soundbay/trainers.py [hunk 5]
```diff
@@ -80,30 +81,39 @@ def train(self):
             if self.debug and epoch > 2:
                 break
 
+
     def train_epoch(self, epoch):
         self.model.train()
         for it, batch in tqdm(enumerate(self.train_dataloader), desc='train'):
-            if it == 3 and self.debug:
+            if it == 2:
```
Comment (mosheman5) [inline soundbay/trainers.py:88]: could have been a funny bug :)
can you please return the `and self.debug`?

File: soundbay/trainers.py [hunk 6]
```diff
@@ -80,30 +81,39 @@ def train(self):
...
                 break
 
             self.model.zero_grad()
-            audio, label = batch
-            audio, label = audio.to(self.device), label.to(self.device)
+            audio, label, raw_wav, idx = batch
+            audio, label, raw_wav, idx  = audio.to(self.device), label.to(self.device), raw_wav.to(self.device), idx.to(self.device)
```
[the remark this answers is not in the document]
Comment (mosheman5) [reply soundbay/trainers.py:93]: agree

File: soundbay/trainers.py [hunk 7]
```diff
@@ -80,30 +81,39 @@ def train(self):
...
 
+            if it == 1:
+                self.logger.upload_artifacts(audio, label, raw_wav, idx, sample_rate=self.train_dataloader.dataset.sample_rate, flag='train', epoch=epoch)
```
Comment (mosheman5) [inline soundbay/trainers.py:96]: could you please verify if works in debug mode as well? I mean, that the code works but nothing is uploaded ofcourse

File: soundbay/utils/logging.py (hunk 1 above)
Comment (mosheman5) [reply soundbay/utils/logging.py:145]: or even normalize it always to -6dB, something like:
`artifact_wav = torch.squeeze(raw_wav).detach().cpu().numpy() / (wav_wav.abs().max()) * 0.5`
the current implementation might introduce clipping in some wavs

File: soundbay/utils/logging.py [hunk 8]
```diff
@@ -125,11 +128,43 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+        """upload algorithm artifacts to W&B during training session"""
+        matplotlib.use('Agg')
+        idx = idx.detach().cpu().numpy()
+        label = label.detach().cpu().numpy()
+
+
+        # Original wavs batch
+
+        artifact_wav = torch.squeeze(raw_wav).detach().cpu().numpy() * 100
+        artifact_wav_list = [artifact_wav[x,...] for x in range(artifact_wav.shape[0])]
```
Comment (mosheman5) [inline soundbay/utils/logging.py:146]: I think you can directly iterate on the numpy object, no need to create a list for that

File: soundbay/utils/logging.py [hunk 9]
```diff
@@ -125,11 +128,43 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+        list_of_wavs_objects = [wandb.Audio(data_or_path=wav, caption=f'label_{lab}_{ind}_train', sample_rate=sample_rate) for wav, ind, lab in zip(artifact_wav_list,idx, label)]
+
+        # Spectrograms batch
+        artifact_spec = torch.squeeze(audio).detach().cpu().numpy()
+        axes = [plt.subplots(nrows=1, ncols=1) for _ in range(artifact_spec.shape[0])]
+        specs = [librosa.display.specshow(artifact_spec[x,...], ax=axes[x][1]) for x in range(artifact_spec.shape[0])]
+        list_of_specs_objects = [wandb.Image(data_or_path=spec, caption=f'{ind}_train') for spec, ind in zip(specs,idx)]
```
Comment (mosheman5) [inline soundbay/utils/logging.py:153]: can you add label to the images as well? as you did in the audio artifact

File: soundbay/utils/logging.py (hunk 3 above)
Comment (mosheman5) [reply soundbay/utils/logging.py:162]: and if all are required it might be useful to have

`def clear_plt_state(): \n
    plt.figure().clear() \n
    plt.close() \n
    plt.cla() \n
    plt.clf()`

but it really a non issue

File: soundbay/utils/logging.py [hunk 10]
```diff
@@ -125,11 +128,43 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+        log_dict1 = {f'First batch {flag} original wavs for epoch {epoch}': list_of_wavs_objects}
```
Comment (mosheman5) [inline soundbay/utils/logging.py:154]: rename `log_dict1` to `wavs_dicts` and `log_dict2` to `specs_dict`/`processed_dict` ?
also, better to remove the epoch from the title, so we could enjoy the scroll bar feature of wandb:
`wavs_dicts = {f'First batch {flag} original wavs': list_of_wavs_objects}`
same for specs

**Review by mosheman5 — CHANGES_REQUESTED** [review]
left some minor comments 
Overall, looks great and it's a super important feature!

File: soundbay/data.py [hunk 11]
```diff
@@ -328,7 +339,7 @@ class InferenceDataset(Dataset):
     '''
     class for storing and loading data.
     '''
-    def __init__(self, file_path: (str, Path),
+    def __init__(self, file_path: Union[str, Path],
```
[the remark this answers is not in the document]
Comment (Z30G0D) [reply soundbay/data.py:342]: Actually it works fine, the linting was annoying xD

File: soundbay/utils/logging.py (hunk 1 above)
Comment (Z30G0D) [reply soundbay/utils/logging.py]: Well it was very confusing.
@noamzbr I think you are missing `log10(1/ np.std(artifact_wav )` in the lower signal expression.
Otherwise I have to put in a negative volume.
Had to catch up with your overwhelming acoustics experience xD

File: soundbay/utils/logging.py [hunk 12]
```diff
@@ -125,11 +129,46 @@ def calc_metrics(self, epoch: int, mode: str = 'train', label_names: List[str] =
...
+
+        artifact_wav_list = [artifact_wav[x,...] for x in range(artifact_wav.shape[0])]
+        list_of_wavs_objects = [wandb.Audio(data_or_path=wav, caption=f'label_{lab}_{ind}_train', sample_rate=sample_rate) for wav, ind, lab in zip(artifact_wav_list,idx, label)]
+
+        # Spectrograms batch
+        artifact_spec = torch.squeeze(audio).detach().cpu().numpy()
+        axes = [plt.subplots(nrows=1, ncols=1) for _ in range(artifact_spec.shape[0])]
+        specs = [librosa.display.specshow(artifact_spec[x,...], ax=axes[x][1]) for x in range(artifact_spec.shape[0])]
+        list_of_specs_objects = [wandb.Image(data_or_path=spec, caption=f'label_{lab}_{ind}_train') for spec, ind, lab in zip(specs,idx, label)]
+        log_wavs = {f'First batch {flag} original wavs for epoch {epoch}': list_of_wavs_objects}
```
Comment (mosheman5) [inline soundbay/utils/logging.py:161]: better to remove the epoch from the title, so we could enjoy the scroll bar feature of wandb:
`wavs_dicts = {f'First batch {flag} original wavs': list_of_wavs_objects}`
same for specs

Comment (Z30G0D) [reply soundbay/utils/logging.py]: re `spces` do you have any suggestion how to make without a list ( I need to plot into all the axes) seperately.
Maybe I am missing out on something?

File: soundbay/trainers.py [hunk 13]
```diff
@@ -80,29 +81,36 @@ def train(self):
...
 
             # estimate and calc losses
             estimated_label = self.model(audio)
             loss = self.criterion(estimated_label, label)
             loss.backward()
             self.optimizer.step()
 
-            # update losses
+            # update losses and log batch
+            # TODO: add logging of batch to logger
```
Comment (mosheman5) [inline soundbay/trainers.py:105]: now this TODO can be removed :D

**Review by mosheman5 — APPROVED** [review]
Very important feature, noice!
